Executors: add native k8s executors

sanderginn opened the pull request:
## Description
This PR adds a chart for deploying native k8s executors. It also polishes the dind implementation.

### Checklist

- [ ] Follow the [manual testing process](https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/TEST.md)
- [ ] Update [changelog](https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/charts/sourcegraph/CHANGELOG.md)

### Test plan
- Ran the test steps as described in https://github.com/sourcegraph/sourcegraph/issues/50824
- Unit tests
<!--
  As part of SOC2/GN-104 and SOC2/GN-105 requirements, all pull requests are REQUIRED to
  provide a "test plan". A test plan is a loose explanation of what you have done or
  implemented to test this, as outlined in our Testing principles and guidelines:
  https://docs.sourcegraph.com/dev/background-information/testing_principles
  Write your test plan here after the "Test plan" header.
-->

sanderginn commented:
Here's an example `override.yaml` that I used to deploy executors with to the test cluster (in this case, a `batches` executor).

I created a dedicated node with the label `sourcegraph-executors=true` in order to target it with the node selector. This is required because the executor spawns a job that mounts the same PVC as the executor deployment, so they must reside on the same node. I will propose the addition of setting pod affinity on the jobs, because the current implementation doesn't support pain-free node autoscaling.

```yaml
executor:
  queueName: batches
  nodeSelector:
    sourcegraph-executors: "true"
  env:
    EXECUTOR_FRONTEND_URL:
      value: "http://sourcegraph-frontend:30080"
    # -- The shared secret configured in the Sourcegraph instance site config under executors.accessToken. Required.
    EXECUTOR_FRONTEND_PASSWORD:
      value: "hunter2hunter2hunter2hunter2"
    # -- The name of the queue to pull jobs from to. Possible values: batches and codeintel. Required.
    EXECUTOR_QUEUE_NAME:
      value: "batches"
    EXECUTOR_MAXIMUM_NUM_JOBS:
      value: "10"
    SRC_LOG_LEVEL:
      value: info
    SRC_LOG_FORMAT:
      value: condensed
    SRC_TRACE_LOG:
      value: "false"
    EXECUTOR_KUBERNETES_RESOURCE_REQUEST_MEMORY:
      value: 1Gi
    # For local deployments the host is 'host.docker.internal' and this needs to be true
    EXECUTOR_DOCKER_ADD_HOST_GATEWAY:
      value: "false"
    # Useful for debugging.
    EXECUTOR_KUBERNETES_NAMESPACE:
      value: sourcegraph
    KUBERNETES_KEEP_JOBS:
      value: "true"
    EXECUTOR_KEEP_WORKSPACES:
      value: "true"
```

Piszmog commented on charts/sourcegraph-executor/dind/values.yaml:
```diff
@@ -52,68 +52,25 @@ storageClass:
 executor:
   enabled: true
   image:
     defaultTag: 5.0.6@sha256:f78a0ba3d84941ec3577f19075e935bec34a9bfaa9dcb85908142fff017e9d6d
```
Is this version controlled for the whole repo? Wondering since `EXECUTOR_QUEUE_NAMES` is introduced in 5.1.x

sanderginn replied:
good catch. Updated to the same 5.1 image as the k8s chart. I think the images get updated during the release of a new chart version

Piszmog commented on charts/sourcegraph-executor/k8s/templates/docker-daemon.ConfigMap.yaml:
We don't need this for native k8s. Since native k8s interacts with the k8s api, executors do not do anything directly with docker registry

Piszmog commented on charts/sourcegraph-executor/k8s/templates/executor.Deployment.yaml:
```diff
@@ -0,0 +1,116 @@
+      {{- end }}
+      volumes:
+        {{- if .Values.privateDockerRegistry.enabled }}
+        - name: docker-config
```
This can be removed

Piszmog commented on charts/sourcegraph-executor/k8s/templates/executor.Deployment.yaml:
```diff
@@ -0,0 +1,116 @@
+            periodSeconds: 5
+          volumeMounts:
+            {{- if .Values.privateDockerRegistry.enabled }}
+            - mountPath: /etc/docker/daemon.json
```
Remove

Piszmog commented on charts/sourcegraph-executor/k8s/templates/executor.Rolebinding.yaml:
Should this be renamed `executor.RoleBinding.yaml` (big `B`)?

Piszmog commented on charts/sourcegraph-executor/k8s/templates/private-docker-registry.Deployment.yaml:
don't need this for k8s

Piszmog commented on charts/sourcegraph-executor/k8s/templates/private-docker-registry.PersistentVolumeClaim.yaml:
same

Piszmog commented on charts/sourcegraph-executor/k8s/templates/private-docker-registry.Service.yaml:
same

Piszmog reviewed (commented):
Overall makes sense (first time I am seeing this syntax). The only change is that we do not need the private docker registry for the native k8s options. Since that doesn't do anything with docker directly but relies on kubernetes to do the creating of pods/containers

danieldides commented on charts/sourcegraph-executor/README.md:
```diff
@@ -1,102 +1,16 @@
-[sourcegraph/sourcegraph]: ../sourcegraph/
-[sourcegraph/sourcegraph-executor]: ./
-[executor]: https://docs.sourcegraph.com/admin/executors
+## dind
```
I think we should prefer "Docker in Docker (dind)" as the title. I'm not sure how common `dind` is for most people.

sanderginn commented:
@Piszmog resolved the issues, I think it's a-okay for merging now

Piszmog commented on charts/sourcegraph-executor/README.md:
```diff
@@ -1,102 +1,16 @@
+This directory contains two Helm charts for deploying executors on Kubernetes. [Read the docs](doc/admin/executors/deploy_executors_kubernetes.md) for more information on executors on Kubernetes.
 
-This chart contains two deployments, Sourcegraph Executors and a private Docker Registry. It is a supplemental chart for the parent [sourcegraph/sourcegraph] Helm Chart if you wish to deploy executors
+## k8s
```
Similar to the `dind`, maybe make this "Native Kubernetes"

Piszmog reviewed (approved):
Just one more minor change

sanderginn commented:
@jac thanks for that catch 👍 
I've added a separate step for packaging the executor charts.  

I've also changed the names of the charts. I've suffixed them with the implementation. This means that the `dind` chart, that used to be published as `sourcegraph-executor`, is now published as `sourcegraph-executor-dind`. Do you think this is an issue, as customers with the installed chart will have to use the new chart name to upgrade? (I don't know if there are many customers with this deployed, if any, tbh). I can revert the dind one to `sourcegraph-executor` if you think that's better.
